manager/manager.go

```diff
@@ -315,29 +336,69 @@ func (m *Manager) issue(ctx context.Context, volumeID string) error {
 		return fmt.Errorf("driver is not ready to request a certificate for this volume: %v", reason)
 	}
 
-	key, err := m.generatePrivateKey(meta)
-	if err != nil {
-		return fmt.Errorf("generating private key: %w", err)
-	}
-	log.V(2).Info("Obtained new private key")
-
 	csrBundle, err := m.generateRequest(meta)
 	if err != nil {
 		return fmt.Errorf("generating certificate signing request: %w", err)
 	}
 	log.V(2).Info("Constructed new CSR")
 
-	csrPEM, err := m.signRequest(meta, key, csrBundle.Request)
-	if err != nil {
-		return fmt.Errorf("signing certificate signing request: %w", err)
+	var req *cmapi.CertificateRequest
+	var key crypto.PrivateKey
+	// If there is a pending request matches the csrBundle, then we don't need to generate a new one.
+	// If encounter any error, remove the request from cache, and we will generate a new request in next step.
+	if reqItem, exist := m.getPendingRequestItemForVolume(volumeID); exist {
+		if pendingRequest, err := m.lister.CertificateRequests(reqItem.Namespace).Get(reqItem.Name); err != nil {
+			log.Error(err, "Failed fetch pending CertificateRequest")
+			m.deletePendingRequestItemForVolume(volumeID)
+		} else {
+			x509Request, err := pki.DecodeX509CertificateRequestBytes(pendingRequest.Spec.Request)
+			if err != nil {
+				log.Error(err, "Failed decode pending CertificateRequest")
+				m.deletePendingRequestItemForVolume(volumeID)
+			}
+			if isX509CertificateRequestSemanticEqual(x509Request, csrBundle.Request) {
+				// if this pending request's parameters match the csrBundle's request, reuse it
+				if err := privateKeyMatchesCSR(reqItem.Key, x509Request); err != nil {
+					log.Error(err, "Cached private key does not match CSR")
+					m.deletePendingRequestItemForVolume(volumeID)
+				} else {
+					req, key = pendingRequest, reqItem.Key
+					log.V(2).Info("Retry pending request",
+						"namespace", reqItem.Namespace, "name", reqItem.Name)
+				}
+			}
+		}
 	}
-	log.V(2).Info("Signed CSR")
+	// Generate new certificate request if there is no matching request in cache
+	if req == nil {
+		if err := m.cleanupStaleRequests(ctx, log, volumeID); err != nil {
```

Comment:
I don't wanna run `cleanupStaleRequests` conditionally either. However I found if `maxRequestsPerVolume == 1`, the pending request will be deleted.

Initially I also thought `<=` is the problem, but after ran a few examples for a normal case, this makes sense to the definition --- "maximum number of CertificateRequests that should exist at any time for each volume". For example, if `maxRequestsPerVolume == 1`, changing condition to `<=` will end up 2 CertificateRequests left after the issue() call.

Reply:
I don't agree with that reasoning and I think it makes this code even harder to reason about. I would expect that the code bit that ensures that no more than maximum CertificateRequests exist in the cluster to actually do exactly that at the time it runs, not to make assumptions about how many CertificateRequests there will be in cluster after the caller has finished running.

If we want it to ensure that there aren't more CertificateRequests than maximum in cluster post-issuance, we could just run the cleanup function at the end of the issue function.

Reply:
Ultimately I think the reason for the conditional part here is that we only want to clean up once we _know_ we do need to create another request (because we use an absolute number based limit). Perhaps we can simplify this though, by permitting N *older* requests, +1 for the ongoing request. So if the maximum history limit is 3, we will allow up to 3 requests in a *terminal state* and one extra for the currently in-flight request.

The only complexity here then comes about if we have 3 old requests and one in-flight, and then *something* changes whilst that existing 1 is in-flight which causes another to be triggered immediately. In those cases we'd likely need to delete the currently in-flight request instead.

So.. to summarise/answer this actual thread, I think we may want to separate out the 'budget' for history from the budget for 'in flight' requests (which IMO should always be 1), so there two parts don't get so intertwined/confusing.. WDYT?

Reply:
Like the idea to define the `budget` for CRs in terminal state. And in-flight should always be one I believe. Shall this be a separate issue to track ?